_posts/2018-06-18-On-Obsessive-Statelessness.markdown

```diff
@@ -0,0 +1,201 @@
+---
+layout: epic
+title: "On the obsessive statelessness of Peril"
+date: 2018-06-18
+author: [orta]
+categories: [danger, peril, state]
+css: danger
+---
+
+We're at 9 months of serious usage of Peril in Artsy. However, I've been worried.
+
+To get you up to speed on Peril, Peril is a tool that takes GitHub webhooks, and makes it easy to build one-off
+actions. It does this by having a per-account settings JSON, that connects JavaScript files to events from webhooks.
+So, for example, you can write a rule which runs when closing an issue in GitHub that looks for associated Jira
+tickets and resolves them. Peril provides no implicit actions like that, it instead offers a JavaScript runtime
+environment optimised to this domain so you can make actions to fit your needs. Like a collection of single-file
+[probots][probots].
+
+Three months ago I started building out a "true" staging environment for Peril, one that allows any user or org on
+GitHub to click a button and have Peril running on their account. Pulling this off has two real interesting
+problems. Problem number one, security. Problem number two, my wallet.
+
+Both of these issues stem from one simple problem: I need to run other people's code on my machines and I think they
+should be able to store data. Which to be quite frank, is horrifying for a side-project. So, this post explores one
+of main aspects which I've architected Peril to make this problem tractable. Avoiding storing state in the form of
+data.
+
+<!-- more -->
+
+## Evaluation Context
+
+Let's start with grounding how Peril works. The GitHub term for when someone adds Peril to their account is that it
+creates an "Installation" of the
+[GitHub App](https://blog.github.com/2016-09-14-a-whole-new-github-universe-announcing-new-tools-forums-and-features/#integrate-seamlessly-with-github).
+When a webhook from GitHub is sent to Peril, Peril grabs the installation ID out a mongo database (yeah, I know,
```

Comment:
(Gonna ask a bunch of questions that come from me not having a lot of hands-on experience with Peril. But maybe that will be helpful in making this accessible to readers with similar lack of background.)

What does "sent to Peril" mean? Is Peril running on a central server?

Reply:
Curious why Peril would need to grab an installation ID out of Mongo. Naively, I would imagine that GitHub webhooks would send an installation ID, which would be used by Peril to pull out the locally stored rules associated with that installation.

You mentioned up above that there is a per-account settings JSON. Is this the same as what's being pulled from Mongo here? I'm just curious, because I think of Mongo as storing data in an abstract sense, and JSON more as a concrete format for serialization to file or network connection. (Even though I know Mongo's storage model is very JSONish.)

Reply:
In Mongo, we have this:

```
{
    "_id" : ObjectId("5adc346471a694013d3bb4a0"),
    "iID" : 135226,
    "login" : "danger",
    "settings" : {
        "modules" : [ 
            "danger-plugin-spellcheck", 
            "danger-plugin-yarn", 
            "@slack/client", 
            "semver-sort"
        ]
    },
    "perilSettingsJSONURL" : "danger/peril-settings@settings.json",
    "__v" : 0,
    "tasks" : {
        "trigger-scheduler" : "danger/peril-settings@testing/trigger-scheduler.ts",
        "logger" : "danger/peril-settings@testing/logger.ts"
    },
    "rules" : {
        "create (ref_type == tag)" : [ 
            "danger/peril-settings@org/new_tag.ts", 
            "danger/peril-settings@org/updateDangerSystems.ts"
        ],
        "pull_request___closed" : "danger/peril-settings@org/aeryn.ts",
        "pull_request" : [ 
            "danger/peril-settings@org/debug.ts", 
            "danger/peril-settings@org/changelog.ts"
        ],
        "issue_comment" : "danger/peril-settings@org/markAsMergeOnGreen.ts",
        "status___success" : "danger/peril-settings@org/mergeOnGreen.ts"
    },
    "repos" : {},
    "recordWebhooksUntilTime" : ISODate("2018-06-16T21:26:48.467Z"),
    "envVars" : {
        "GITLAB_DEPLOY_TOKEN" : "12345"
    },
    "startedRecordingWebhooksTime" : ISODate("2018-06-16T21:21:48.467Z")
}
```

It's basically [the user's JSON](https://github.com/danger/peril-settings/blob/master/settings.json), env vars, and mixed with some unique to Peril data. I expanded the definition